### PR TITLE
Normalize lesson keys

### PR DESCRIPTION
**Pull request by Hamms**
Specifically, convert from the Lesson URLs that we use as keys in Crowdin to something that's going to be easier to programmatically reference when rendering translations.

Depends on https://github.com/code-dot-org/code-dot-org/pull/41003

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/lib/services/globally_unique_identifiers.rb
Comment (davidsbailey) [inline dashboard/lib/services/globally_unique_identifiers.rb:89]: Good call out! I don't totally follow your comment, though. It looks like the reason you can't implement the inverse operation is because the delimiter is a slash, but lesson keys are allowed to have slashes in them, is that correct? If so, it doesn't seem like that would change if we added validation.

Do you anticipate needing to do the inverse operation in the future? If so, what do you think about using a different delimiter character when joining below?

Comment (Hamms) [reply dashboard/lib/services/globally_unique_identifiers.rb:89]: I'd say it's a slightly broader issue than that; it's not just that we're using a slash as a delimiter and are worried about slashes showing up in the keys, it's that we don't enforce ANYTHING about the format of the keys, so no matter WHAT we use as a delimiter there's a chance of an issue. Similar to how we had to add validation for vocab, resource, and course version in order to get the special markdown syntaxes to work.

I don't currently anticipate needing to do the inverse operation. If we did end up wanting it, I think it would just be a matter of enforcing formatting for these model values (which is something that would be good to do anyway) and then picking a delimiter that works with that enforced format.

File: bin/i18n/curriculum_sync_utils.rb
Comment (bethanyaconnor) [inline bin/i18n/curriculum_sync_utils.rb]: I don't think this is a safe way to look up lessons. It is possible for multiple lessons in a script to have the same `relative_position` because of lockable lessons. For example, locally I get:
```
[development] dashboard > Lesson.joins(:script).where("scripts.name": 'csd1-2021', relative_position: 1).count
   (3.1ms)  SELECT COUNT(*) FROM `stages` INNER JOIN `scripts` ON `scripts`.`id` = `stages`.`script_id` WHERE `stages`.`relative_position` = 1 AND `scripts`.`name` = 'csd1-2021'
=> 2
```

I think just querying for lessons that have `has_lesson_plan` set to true will fix this

File: dashboard/lib/services/globally_unique_identifiers.rb
Comment (bethanyaconnor) [inline dashboard/lib/services/globally_unique_identifiers.rb]: nit: s/Resource/Lesson

**Review by davidsbailey — APPROVED** [review]
LGTM after addressing Bethany's comments.

File: bin/i18n/curriculum_sync_utils.rb
Comment (Hamms) [reply bin/i18n/curriculum_sync_utils.rb]: Ha, I actually went and looked up https://github.com/code-dot-org/code-dot-org/blob/653b730a8505385ec1085a5d4cd6d085c8b4348f/dashboard/app/controllers/lessons_controller.rb#L24-L26 to see what the right way to do this was; I noticed that it checked `has_lesson_plan` and made a mental note to figure out _why_ so I could know whether or not it needed to be included here, and then I just didn't. Thanks for catching!

**Review by daynew — APPROVED** [review]
Looks from a i18n sync perspective. I'm not familiar with the database models though, so make sure to get signoff from an expert.

File: dashboard/app/models/lesson.rb
Comment (bethanyaconnor) [inline dashboard/app/models/lesson.rb:74]: @Hamms you might already know this but this is the full explanation for why there can be two lessons with the same `relative_position`. I don't think we would ever translate anything that didn't have `has_lesson_plan` set to true but it might be worth querying for `!lockable || has_lesson_plan` in the query above

**Review by bethanyaconnor — APPROVED** [review]

**Pull request merged**
